### DISCUSSION TITLE
How to use Rubric Endpoint to update grade of a specific assignment's Rubric Assesment

### DISCUSSION THREAD

**Discussion by Volan360** [ucfopen/canvasapi#617] [Q&A] [answered]
Hi, I was digging around and noticed some resolved issues showing that an endpoint has been made for all the Rubric functionality in the Canvas API. I was wondering if I could get some help actually using this endpoint?

At the moment I can get the Assignment ID (and all submissions to that assignment) and the relevant Rubric ID, but I'm not sure how to get the actual Rubric Assessment that each submission is currently using, so I can update the columns correctly.

For context, the assignments are already linked to Rubrics on Canvas, and there have already been submissions before. I don't want to make a new Rubric Assessment, I'd like to update the scores of the existing submissions' Rubric assessment, specifically the points value of rubric item "Fundamental Skills 1"

Using the documentation at https://canvas.instructure.com/doc/api/submissions.html#method.submissions_api.update to find the relevant kwargs, this is where I'm at code-wise. I get the error: `canvasapi.exceptions.BadRequest: {"message":"invalid rubric_assessment"}`, which I assume means I structured the rubric_assessment argument wrong
``` 
canvas = Canvas(API_URL, API_KEY)
course = canvas.get_course(COURSE_ID)
assignment = course.get_assignment(ASSIGNMENT_ID)
submissions = assignment.get_submissions()
for submission in submissions:
     res = submissions[0].edit(rubric_assessment={"Fundamental Skills 1": {"points": 2}})
```

Thanks for this wonderful library!

**Comment by Thetwam** [comment]
Hey @Volan360,

At a glance, your code looks mostly correct. The only weird thing about the `.edit` call is that `rubric_assessment` appears to be looking for a Rubric Criterion ID. `"Fundamental Skills 1"` looks more like a title or description to me. Do you have an ID you can try as that key instead?

**Reply by bennettscience** [reply, marked as answer]
Adding to @Thetwam, to get the rubric assessments, you need to use the `include=['rubric_assessments']` keyword when you're getting your submissions.

```python
submissions = assignment.get_submissions(include=['rubric_assessments'])
submissions[0].rubric_assessment

"""
{ 
    '_1234': {
        'rating_id': '_5678',
        'comments': 'Some string',
        'points': 5.0
    },
    { ...},
}
"""
```

The outermost ID is the `RubricAssociation` ID you can use to get the rubric itself if you want to. Each object in the association is a row on the rubric, so you might also need some logic to check that you're updating the correct row. 

Rubrics are tricky in Canvas because they're nested so deeply. Criterion are nested on the rubric, which are nested on the assignment. Results are nested for each student under each assignment...it can get deep quicky.

I would refactor what you have to work on an assignment with a single-criterion rubric to get your login in the right place. Then you can add one more check to handle rubrics with multiple criteria to grade.

**Reply by Volan360** [reply]
Thanks for the help both of you. I'll take your suggestion into consideration, but the info about including Rubric Assessments in the submission query to get the IDs should solve my issue for now!

Edit: Should be include=['rubric_assessment'] rather than 'rubric_assessments' by the way, but other than that this solution was perfect! I get the relevant rubric from assignment object and find the correct criterion by checking against the descriptions.
